CoseSign1/DetachedSignatureFactory.cs

```diff
@@ -228,22 +377,57 @@ public DetachedSignatureFactory(HashAlgorithmName hashAlgorithmName, ICoseSign1M
             throw new ArgumentNullException("payload", "Either streamPayload or bytePayload must be specified, but not both at the same time, or both cannot be null");
         }
 
-        ReadOnlyMemory<byte> hash = streamPayload != null
-                                    ? InternalHashAlgorithm.ComputeHash(streamPayload)
-                                    : InternalHashAlgorithm.ComputeHash(bytePayload!.Value.ToArray());
+        ReadOnlyMemory<byte> hash;
+        string extendedContentType;
+        if (!payloadHashed)
+        {
+            hash = streamPayload != null
+                                 ? InternalHashAlgorithm.ComputeHash(streamPayload)
+                                 : InternalHashAlgorithm.ComputeHash(bytePayload!.Value.ToArray());
+            extendedContentType = ExtendContentType(contentType);
+        } else
+        {
+            hash = streamPayload != null
+                                 ? streamPayload.GetBytes()
+                                 : bytePayload!.Value.ToArray();
+            try
+            {
+                HashAlgorithmName algoName = SizeToAlgorithm[hash.Length];
+                extendedContentType = ExtendContentType(contentType, algoName);
+            }
+            catch (KeyNotFoundException e)
+            {
+                throw new ArgumentException($"{nameof(payloadHashed)} is set, but payload size does not correspond to any known hash sizes in {nameof(HashAlgorithmName)}", e);
+            }
+        }
+
+
         return returnBytes
                ? InternalMessageFactory.CreateCoseSign1MessageBytes(
                     hash,
                     signingKeyProvider,
                     embedPayload: true,
-                    contentType: ExtendContentType(contentType))
+                    contentType: extendedContentType)
                : InternalMessageFactory.CreateCoseSign1Message(
                     hash,
                     signingKeyProvider,
                     embedPayload: true,
-                    contentType: ExtendContentType(contentType));
+                    contentType: extendedContentType);
     }
 
+    /// <summary>
+    /// quick lookup of algorithm name based on hash size
+    /// </summary>
+    private static readonly ConcurrentDictionary<int, HashAlgorithmName> SizeToAlgorithm = new(

```

Comment:
Recommend adding the word Bytes to this and update the comment so maybe something like SizeInBytes...